feat: add YAML input format support to wdl-engine

Gyan-max opened the pull request:
<!-- START SECTION: any other pull request -->

This PR adds YAML support for input files in the WDL engine. It allows users to provide input files in YAML format (.yml or .yaml) alongside the existing JSON format.

Specifically, this PR:
1. Adds the `serde_yaml` dependency to the WDL engine
2. Implements an `InputFormat` trait with implementations for both JSON and YAML
3. Extends the `Inputs` struct with `parse_json` and `parse_yaml` methods
4. Updates the test framework to test both formats
5. Adds test cases for YAML input parsing

Before submitting this PR, please make sure:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [x] You have added at least one relevant code reviewer to the PR.
- [x] Your code builds clean without any errors or warnings.
- [x] You have added tests (when appropriate).
- [x] You have updated the README or other documentation to account for these
      changes (when appropriate).
- [x] You have added an entry to the relevant `CHANGELOG.md` (see
      ["keep a changelog"] for more information).
- [x] Your commit messages follow the [conventional commit] style.
- [x] Your PR title follows the [conventional commit] style.

<!-- END SECTION -->

[conventional commit]: https://www.conventionalcommits.org/en/v1.0.0/#summary
["keep a changelog"]: https://keepachangelog.com/en/1.1.0/

peterhuene commented:
Hi @Gyan-max, thanks for this PR! I should be able to get it reviewed tomorrow morning (Eastern Daylight Savings time).

Gyan-max commented:
> Hi @Gyan-max, thanks for this PR! I should be able to get it reviewed tomorrow morning (Eastern Daylight Savings time).

sure
and could you please provide guidance on the next steps required for the '[feat: YAML input support](https://github.com/stjude-rust-labs/sprocket/pull/69)' pull request (#69)?

peterhuene commented on wdl-engine/Cargo.toml:
```diff
@@ -18,6 +18,7 @@ anyhow = { workspace = true }
 ordered-float = { workspace = true }
 indexmap = { workspace = true }
 serde_json = { workspace = true }
+serde_yaml = "0.9.34"
```
As mentioned in the Sprocket PR, we should be using `serde-yaml-ng` as `serde-yaml` is unmaintained; we have a CI check in this repository that ensure we don't use unmaintained dependencies, so if we want these changes through CI it would fail.

peterhuene replied:
So the `deny` CI action is passing because apparently `serde-yaml` did not issue an "unmaintained" security advisory from what I can tell.

Still, we should no longer be using it as they still might publish a security advisory in the future.

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -27,6 +28,90 @@ use crate::Value;
+///
+/// This trait defines an interface for parsing different file formats into a common
+/// representation (JsonMap) that can be used by the inputs system.
+pub(crate) trait InputFormat {
```
This trait isn't currently necessary as you're not using it in any generic bounds (i.e. you're calling the methods on `JsonInputFormat` and `YamlInputFormat` directly rather than via a generic parameter).

In the Sprocket PR, I mentioned creating a few abstraction traits for passing *both* YAML types and JSON types down into `parse_object` via a generic parameter so that we don't have to convert YAML to JSON.

Your changes here use `serde_json::to_value` to deserialize an already-deserialized YAML value into a JSON value. Doing it this way has a marginal runtime cost for the extra deserialization, but it definitely cuts down on code complexity to abstract using both YAML and JSON types.

Let's do it the way you have in this PR, so I think we can just remove this trait and move the implementations directly into `Inputs::parse_json` and `Inputs::parse_yaml`.

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -27,6 +28,90 @@ use crate::Value;
+        
+        // Convert YAML to JSON format
+        let mut json = serde_json::to_value(yaml)
+            .with_context(|| format!("failed to convert YAML to JSON for processing"))?;
```
There's two issues here: a use of `format!` with no interpolations (i.e. `{}`) is unnecessary and `with_context` is unnecessary if you want to use an unformatted string; use `context` instead:

```suggestion
            .context("failed to convert YAML to JSON for processing")?;
```

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -607,26 +696,46 @@ impl Inputs {
+        
+        // Determine the format based on file extension
+        let object = match path.extension().and_then(|ext| ext.to_str()) {
+            Some("json") => JsonInputFormat::parse_file(path)?,
+            Some("yml") | Some("yaml") => YamlInputFormat::parse_file(path)?,
```
I think these can just be `Self::parse_json` and `Self::parse_yaml` once we remove `JsonInputFormat` and `YamlInputFormat` and move their implementations into `Self::parse_json` and `Self::parse_yaml`, respectively.

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -607,26 +696,46 @@ impl Inputs {
+        let object = match path.extension().and_then(|ext| ext.to_str()) {
+            Some("json") => JsonInputFormat::parse_file(path)?,
+            Some("yml") | Some("yaml") => YamlInputFormat::parse_file(path)?,
+            _ => bail!("Unsupported input file format. Supported formats: .json, .yml, .yaml"),
```
Error messages in our library do not capitalize sentences or end with periods as typically errors can be (optionally) concatenated together for display.

```suggestion
            _ => bail!("unsupported input file format: supported formats are JSON (extension `.json`) and YAML (extensions `.yaml` and `.yml`)"),
```

peterhuene commented on wdl-engine/tests/inputs.rs:
```diff
@@ -138,38 +139,63 @@ fn run_test(test: &Path, result: AnalysisResult, ntests: &AtomicUsize) -> Result
-                }
+    
+    // Define the parsing methods to try
+    let json_path = test.join("inputs.json");
```
I think I would rather have this be a loop over the input file names and require that every inputs test also have a `inputs.yaml` so that we can be assured that the output of every test is the same regardless of using JSON or YAML.

So something like:

```rust
for path in [test.join("inputs.json"), test.join("inputs.yaml")] {
    let result = match Inputs::parse(document, path) {
        Ok(Some((name, inputs))) => match inputs {
            Inputs::Task(inputs) => {
                match inputs
                    .validate(
                        document,
                        document
                            .task_by_name(&name)
                            .expect("task should be present"),
                        None,
                    )
                    .with_context(|| format!("failed to validate the inputs to task `{name}`"))
                {
                    Ok(()) => String::new(),
                    Err(e) => format!("{e:?}"),
                }
            }
            Inputs::Workflow(inputs) => {
                let workflow = document.workflow().expect("workflow should be present");
                match inputs.validate(document, workflow, None).with_context(|| {
                    format!(
                        "failed to validate the inputs to workflow `{workflow}`",
                        workflow = workflow.name()
                    )
                }) {
                    Ok(()) => String::new(),
                    Err(e) => format!("{e:?}"),
                }
            }
        },
        Ok(None) => String::new(),
        Err(e) => format!("{e:?}"),
    };

    let output = test.join("error.txt");
    compare_result(&output, &result)?;
}
```

We'll also need to normalize any reference to `inputs.yaml`, so have something like `s.replace("inputs.yaml", "inputs.json")` in the `normalize` function. That way we shouldn't have to update any existing test baselines.

Then for the tests we simply add a corresponding `inputs.yaml` to each existing test case.

peterhuene commented on wdl-engine/tests/inputs/yaml-only/error.txt:
```diff
@@ -0,0 +1,5 @@
+Failed to parse input file `tests/inputs/yaml-only/inputs.yaml`: failed to parse input file `tests/inputs/yaml-only/inputs.yaml`
```
This error message shouldn't be present if all we're doing is testing yaml inputs.

At any rate, I think we can delete both the `yaml-only` and `yaml-support` test cases as they're not adding anything new to the test matrix; if instead we do the above and add `inputs.yaml` to each test case, we'll ensure that the existing tests all work with YAML inputs too.

peterhuene reviewed (changes requested):
Hi @Gyan-max. This is looking good!

I have some comments and requested changes to address, but I think what you have here in on track!

Gyan-max commented:
@peterhuene i'll try to address all the changes ASAP!

Gyan-max commented:
Hi @peterhuene 

I've implemented the YAML input parsing refactoring we discussed. Here's what I've done:

Simplified the parsing architecture by removing the InputFormat trait and its implementations, moving the functionality directly into Inputs::parse_json and Inputs::parse_yaml methods.
Fixed the package name in Cargo.toml from serde-yaml-ng to serde_yaml_ng and updated all relevant imports.
Improved test coverage by:
Creating YAML versions of all JSON test input files
Refactoring the test framework to validate both formats consistently
Removing the redundant yaml-only and yaml-support test directories
Modified the test runner to prioritize JSON files for compatibility with existing error baselines, ensuring we don't have to rewrite all test expectations.
All tests are now passing. The changes make the codebase more maintainable while ensuring both input formats are thoroughly tested with the same test cases. Let me know if you have any questions or if you'd like me to make any adjustments to this approach!

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -1,5 +1,3 @@
-//! Implementation of workflow and task inputs.
```
We should add this comment back to make CI happy.

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -616,18 +646,50 @@ impl Inputs {
-                .with_context(|| {
-                    format!("failed to parse input file `{path}`", path = path.display())
-                })?
+                .context("failed to parse input file")?
```
For this change to call `context` instead of `with_context`, we're losing the path to the file, which is always useful to have in errors.

Can we change this back to what we had with `with_context`?

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -616,18 +646,50 @@ impl Inputs {
-                        path = path.display()
-                    )
-                })?,
+                .context("expected input file to contain a JSON object")?,
```
Similar to the above comment, I think we should keep the `with_context` call here.

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -616,18 +646,50 @@ impl Inputs {
+                .context("expected input file to contain a JSON object")?,
         );
 
+        Self::parse_object(document, object).context("failed to parse input file")
```
Similar to the above comments, I think we should include the file path in this error.

```suggestion
        Self::parse_object(document, object).with_context(|| format!("failed to parse input file `{path}`", path = path.display()))
```

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -753,13 +815,12 @@ impl Inputs {
                     inputs
                         .set_path_value(document, task, remainder, value)
-                        .with_context(|| format!("invalid input key `{key}`"))?;
+                        .context("invalid input key")?;
```
We should revert the change to this line.

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -753,13 +815,12 @@ impl Inputs {
         for (key, value) in object {
-            let value = serde_json::from_value(value)
-                .with_context(|| format!("invalid input key `{key}`"))?;
+            let value = serde_json::from_value(value).context("invalid input key")?;
```
We should revert the change to this line. When I mentioned using `context` instead of `with_context`, it only applied to the case where we are not formatting a string. This message formats with the invalid input key, which is useful for users.

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -781,13 +842,12 @@ impl Inputs {
                     inputs
                         .set_path_value(document, workflow, remainder, value)
-                        .with_context(|| format!("invalid input key `{key}`"))?;
+                        .context("invalid input key")?;
```
We should revert the change to this line.

peterhuene commented on wdl-engine/tests/inputs.rs:
```diff
@@ -77,10 +78,19 @@ fn normalize(s: &str) -> String {
+    // Normalize any YAML-specific error messages to match JSON format
+    let s = s.replace("failed to parse input file", "failed to parse input file");
+
+    // Remove any YAML-specific error details that might differ from JSON
```
```suggestion
```

peterhuene commented on wdl-engine/tests/inputs.rs:
```diff
@@ -77,10 +78,19 @@ fn normalize(s: &str) -> String {
+    // Normalize references to YAML files to match JSON baselines
+    let s = s.replace("inputs.yaml", "inputs.json");
+
+    // Normalize any YAML-specific error messages to match JSON format
+    let s = s.replace("failed to parse input file", "failed to parse input file");
```
Upon first glance, this replacement seems unnecessary?

```suggestion
```

Gyan-max replied:
you are right, the string was replaced by its own

peterhuene commented on wdl-engine/tests/inputs/explicit-nested-inputs-not-allowed/error.txt:
```diff
@@ -1,5 +1,5 @@
-failed to parse input file `tests/inputs/explicit-nested-inputs-not-allowed/inputs.json`
```
With the expected implementation for this PR, we should not see any changes to test baselines.

I would expect this and every other changes to `error.txt` to not be present.

Gyan-max replied:
All the error.txt files have now been restored to their original state.

peterhuene commented on wdl-engine/src/inputs.rs:
```diff
@@ -616,18 +646,50 @@ impl Inputs {
+
+        // Convert YAML to JSON format
+        let mut json =
+            serde_json::to_value(yaml).context("failed to convert YAML to JSON for processing")?;
```
Sorry about this, as I think this is the line I told you to change to use `context` in the first place, but I previously failed to notice that this doesn't include the file path either. I think we should change it back to `with_context`, but include the path.

```suggestion
            serde_json::to_value(yaml).with_context(|| format!("failed to convert input file `{path}` to JSON for processing", path = path.display()))?;
```

peterhuene reviewed (commented):
This is looking great! Just a few follow-up comments, mostly related to `with_context` vs `context`.

Gyan-max commented:
@peterhuene Thank you for your detailed review! I've made the following changes based on your feedback:

Reverted .context() calls to .with_context():
Changed all .context("invalid input key")? back to .with_context(|| format!("invalid input key {key}"))? for task and workflow inputs. This keeps the error messages clear by showing the specific key that caused the issue.

Added file paths in error messages:
Kept the .with_context() change for YAML to JSON conversion, so the file path is included in error messages.

Restored test baselines:
Reverted all changes to error.txt files to keep the test results consistent.

Removed unnecessary code:
Deleted an extra string replacement line in the normalize function since it wasn’t needed.

peterhuene reviewed (commented):
Hi @Gyan-max. 

I think these changes are really close to being ready.  I just have two remaining comments to address.

claymcleod commented:
Hi @Gyan-max, there are some CI failures if you could look at addressing those.

peterhuene commented:
Weird, github seems to have not submitted the comments I made as part of the review. Let me recomment.

peterhuene commented on wdl-engine/tests/inputs.rs:
```diff
@@ -78,10 +79,58 @@ fn normalize(s: &str) -> String {
+    
+    // Handle "failed to parse input file" with file path
+    if result.starts_with("failed to parse input file `") {
+        result = result.replace(
+            &result.lines().next().unwrap_or(""),
+            "failed to parse input file"
+        );
+    }
+    
+    // Handle "invalid input key" with specific key
+    let lines: Vec<&str> = result.lines().collect();
+    let mut normalized_lines: Vec<String> = Vec::new();
+    
+    for line in lines {
+        if line.contains("invalid input key `") {
+            // Preserve only the "invalid input key" part without the specific key
+            if let Some(idx) = line.find("invalid input key `") {
+                let base = &line[0..idx + "invalid input key".len()];
+                
+                // If line contains additional info after the key, preserve it
+                if let Some(suffix_idx) = line.rfind("`: ") {
+                    let suffix = &line[suffix_idx + 2..]; // +2 to skip "`: "
+                    normalized_lines.push(format!("{}: {}", base, suffix));
+                } else {
+                    normalized_lines.push(base.to_string());
+                }
+            } else {
+                normalized_lines.push(line.to_string());
+            }
+        } else {
+            normalized_lines.push(line.to_string());
+        }
+    }
+    
+    result = normalized_lines.join("\n");
+    
+    // Ensure correct spacing between error parts
+    if result.contains("failed to parse input file\nCaused by:") {
+        result = result.replace("failed to parse input file\nCaused by:", "failed to parse input file\n\nCaused by:");
+    }
```
I don't think this is necessary. If we have updated the error message formats thanks to changes in `with_context`, we should just update the test baselines with `BLESS=1` rather than tryint to normalize them.

Normalizing should really just be for differences in the operating systems we test on, but also for normalizing the inputs file name for these particular tests.

peterhuene commented on wdl-engine/tests/inputs.rs:
```diff
@@ -139,7 +188,24 @@ fn run_test(test: &Path, result: AnalysisResult, ntests: &AtomicUsize) -> Result
+    } else if yaml_path.exists() {
+        yaml_path
+    } else {
+        // If no input file exists, we'll still validate the document
+        // This handles cases where we just want to verify the document parses correctly
+        ntests.fetch_add(1, Ordering::SeqCst);
+        return Ok(());
```
Personally, I think it should be an error if we have neither file present.

I would also like us to require _both_ files for the tests so that we ensure that for every inputs test, both JSON and YAML formats produce no different in output (excepting the input file name).

Earlier I suggested that we simply do:

```rust
for file in ["inputs.json", "inputs.yaml"] {
  let result = match Inputs::parse(document, test.join(file)) {
      Ok(Some((name, inputs))) => match inputs {
          Inputs::Task(inputs) => {
              match inputs
                  .validate(
                      document,
                      document
                          .task_by_name(&name)
                          .expect("task should be present"),
                      None,
                  )
                  .with_context(|| format!("failed to validate the inputs to task `{name}`"))
              {
                  Ok(()) => String::new(),
                  Err(e) => format!("{e:?}"),
              }
          }
          Inputs::Workflow(inputs) => {
              let workflow = document.workflow().expect("workflow should be present");
              match inputs.validate(document, workflow, None).with_context(|| {
                  format!(
                      "failed to validate the inputs to workflow `{workflow}`",
                      workflow = workflow.name()
                  )
              }) {
                  Ok(()) => String::new(),
                  Err(e) => format!("{e:?}"),
              }
          }
      },
      Ok(None) => String::new(),
      Err(e) => format!("{e:?}"),
  };
  
  let output = test.join("error.txt");
  compare_result(&output, &result)?;
}
```

Gyan-max commented:
@peterhuene @claymcleod ThnakYou for the reviews, I'll try to address all the feedback ASAP.

a-frantz commented on wdl-engine/tests/inputs.rs:
```diff
@@ -139,43 +145,55 @@ fn run_test(test: &Path, result: AnalysisResult, ntests: &AtomicUsize) -> Result
+    let json_path = test.join("inputs.json");
+    let yaml_path = test.join("inputs.yaml");
+
+    // Check if at least one of the input files exists
```
Should we require both YAML and JSON be present? I suppose if we start adding other formats that might get annoying and difficult to maintain, but it seems preferable over having blind spots in our test coverage.

peterhuene replied:
I specifically requested adding both formats to the tests as 1) there aren't that many of these tests and 2) it ensures the same result regardless of input format.

I don't see us adding more formats for the foreseeable future, so I'd advocate to just keeping as-is in the PR.

Gyan-max commented:
@a-frantz 

1. Updated the test code in `inputs.rs` to require both JSON and YAML input files for all tests except the "missing-file" test that specifically tests file absence.
2. Updated the file documentation to clearly state that both input formats are required for all tests, ensuring complete coverage across different formats.
3. Modified the error messages to be clear about the requirement for both formats.
4. The changes ensure that all test cases thoroughly validate both JSON and YAML input formats with the same test data, preventing blind spots in the test coverage.

Gyan-max commented:
Hi @a-frantz 
I'm confused what is the conflict in this branch. I checked my local repo and I don't see any merge conflict in the CHANGELOG.md file and I'm not getting any clear insight into what exactly needed to be resolved form the github interface's message.
Could you please help me out.

a-frantz commented:
> Hi @a-frantz I'm confused what is the conflict in this branch. I checked my local repo and I don't see any merge conflict in the CHANGELOG.md file and I'm not getting any clear insight into what exactly needed to be resolved form the github interface's message. Could you please help me out.

I made things worse 😅 The conflict was very simple, and we wanted to just accept both changes to the CHANGELOG (the conflict was caused by `main` and your branch adding entries at the same location. We want both entries). But then I was blocked from forcepushing your branch. I then tried merging my force push with your branch, and goofed it up pretty bad. I can't fix this easily without force push permissions.

I can walk you through the commands to fix this if you message me on slack! But I can't do anything more remotely without making it worse. A force push is going to be required.

claymcleod commented:
Can you please address the outstanding CI issue?

Gyan-max commented:
> Can you please address the outstanding CI issue?

Hey @claymcleod Could you please help me out in this.
I'm getting little bit confused

peterhuene reviewed (approved):
Hi @Gyan-max.

This looks good! I pushed up a squash of your commits and rebased it on `main` to resolve merge conflicts.

Should be ready to merge once CI passes.

a-frantz reviewed (approved):
LGTM 👍

peterhuene commented:
@Gyan-max Thank you for the contribution!
